tests/test_send_by_kwargs.py

```diff
@@ -0,0 +1,168 @@
+    )
+    @pytest.mark.parametrize(
+        argnames='timeout', argvalues=[None, 5], ids=['default_timeout', 'custom_timeout']
+    )
+    def test_defaults_handling(self, method, timeout, bot, monkeypatch):
+        def make_assertion(**kwargs):
+            if timeout is None:
+                self.test_flag = 'timeout' not in kwargs
+            else:
+                self.test_flag = kwargs.get('timeout', None) == 5
+
+        signature = inspect.signature(getattr(bot, method))
+        kwargs = {
+            name: True
+            for name, param in signature.parameters.items()
+            if param.default == inspect.Parameter.empty
+            # special casing for some methods where the required arguments are not clear
+            # due to the fact that we made them optional in order to allow passing e.g. a Venue
+            # directly
+            or name in ['latitude', 'longitude', 'address', 'phone_number', 'first_name']
+        }
+        if timeout is not None:
+            kwargs['timeout'] = 5
+
+        # we're a bit tricky here, because otherwise monkeypatch would fiddle with the signature
+        CACHED_SIGNATURES['make_assertion'] = signature
+
+        monkeypatch.setattr(bot, method, make_assertion)
+        send_by_kwargs(bot, kwargs)
+        assert self.test_flag
+
+        self.test_flag = False
+        send_by_kwargs(bot, **kwargs)
+        assert self.test_flag
+
+    def test_fallback_behavior(self, monkeypatch, bot):
+        def make_assertion(**kwargs):
+            self.test_flag = True
+
+        signature = inspect.signature(bot.send_dice)
+        CACHED_SIGNATURES['make_assertion'] = signature
+        monkeypatch.setattr(bot, 'send_dice', make_assertion)
+        send_by_kwargs(bot, {'chat_id': 1})
+        assert self.test_flag
+
+        with pytest.raises(RuntimeError, match='Could not find a bot method'):
+            send_by_kwargs(bot, {})
+
+    @pytest.mark.parametrize(
+        'kwargs,_kwargs',
+        [
+            [{'chat_id': 123, 'text': 'Hello there'}, {}],
+            [{}, {'chat_id': 123, 'text': 'Hello there'}],
+            [{'chat_id': 123}, {'text': 'Hello there'}],
+            [
+                {'chat_id': 456, 'text': 'General Kenobi'},
+                {'chat_id': 123, 'text': 'Hello there'},
+            ],
+        ],
+    )
+    def test_kwarg_mixing(self, bot, monkeypatch, kwargs, _kwargs):
+        expected_kwargs = {'chat_id': 123, 'text': 'Hello there'}
+
+        def make_assertion(**kw):
+            self.test_flag = kw == expected_kwargs
+
+        signature = inspect.signature(bot.send_message)
+        CACHED_SIGNATURES['make_assertion'] = signature
+        monkeypatch.setattr(bot, 'send_message', make_assertion)
+        send_by_kwargs(bot, kwargs, **_kwargs)
+        assert self.test_flag
+
+    def test_method_raises_exception(self, bot, monkeypatch):
+        def mock(**kw):
+            raise TelegramError()
```

Comment:
Doesn't this take `message` as a compulsory argument? or ig it doesn't matter here since this is being exception handled anyway..

Reply:
yes, I just put `**kw`, b/c I didn't care about which kwargs are actually needed …